src/libguac/encode-webp.c

```diff
@@ -0,0 +1,243 @@
+            writer->buffer, writer->buffer_size);
+
+    /* Clear buffer */
+    writer->buffer_size = 0;
+
+}
+
+/**
+ * Configures the given stream writer object to use the given Guacamole stream
+ * object for WebP output.
+ *
+ * @param writer
+ *     The Guacamole WebP stream writer structure to configure.
+ *
+ * @param socket
+ *     The Guacamole socket to use when sending blob instructions.
+ *
+ * @param stream
+ *     The stream over which WebP-encoded blobs of image data should be sent.
+ */
+static void guac_webp_stream_writer_init(guac_webp_stream_writer* writer,
+        guac_socket* socket, guac_stream* stream) {
+
+    writer->buffer_size = 0;
+
+    /* Store Guacamole-specific objects */
+    writer->socket = socket;
+    writer->stream = stream;
+
+}
+
+/**
+ * WebP output function which appends the given WebP data to the internal
+ * buffer of the Guacamole stream writer structure, automatically flushing the
+ * writer as necessary.
+ *
+ * @param data
+ *     The segment of data to write.
+ *
+ * @param data_size
+ *     The size of segment of data to write.
+ *
+ * @param picture
+ *     The WebP picture associated with this write operation. Provides access to
+ *     picture->custom_ptr which contains the Guacamole stream writer structure.
+ *
+ * @return
+ *     True if writing was successful, false on failure.
+ */
+static int guac_webp_stream_write(const uint8_t* data, size_t data_size,
+        const WebPPicture* picture) {
+
+    guac_webp_stream_writer* const writer = (guac_webp_stream_writer*)picture->custom_ptr;
+    if (writer == NULL) {
+        return 0;
+    }
+
+    const unsigned char* current = data;
+    int length = data_size;
+
+    /* Append all data given */
+    while (length > 0) {
+
+        /* Calculate space remaining */
+        int remaining = sizeof(writer->buffer) - writer->buffer_size;
+
+        /* If no space remains, flush buffer to make room */
+        if (remaining == 0) {
+            guac_webp_flush_data(writer);
+            remaining = sizeof(writer->buffer);
+        }
+
+        /* Calculate size of next block of data to append */
+        int block_size = remaining;
+        if (block_size > length)
+            block_size = length;
+
+        /* Append block */
+        memcpy(writer->buffer + writer->buffer_size,
+               current, block_size);
+
+        /* Next block */
+        current += block_size;
+        writer->buffer_size += block_size;
+        length -= block_size;
+
+    }
+
+    return 1;
+}
+
+int guac_webp_write(guac_socket* socket, guac_stream* stream,
+        cairo_surface_t* surface, int quality) {
+
+    int width = cairo_image_surface_get_width(surface);
+    int height = cairo_image_surface_get_height(surface);
+    int stride = cairo_image_surface_get_stride(surface);
+    cairo_format_t format = cairo_image_surface_get_format(surface);
+
+    if (format != CAIRO_FORMAT_RGB24 && format != CAIRO_FORMAT_ARGB32) {
```

Comment:
Since we cannot support ARGB32 for JPEG, and the JPEG vs. WEBP decision is only made magically within `guac_client_stream_lossy_image()`, the only way this can be used is if code that *shouldn't* be aware of the status of WEBP support within guacamole-server makes assumptions.

If we want to support ARGB for WEBP, I think the only reasonable way would be to:

1. Restore `guac_client_stream_jpeg()`
2. Add `guac_client_stream_webp()`, which always flags an error if WEBP is unsupported by guacamole-server.
3. Add, for convenience, `guac_client_supports_webp()`, which verifies client support for `image/webp`, and simply returns `0` in all cases if WEBP is unsupported by guacamole-server.

What do you think?